src/core/reactor_backend.cc

```diff
@@ -1484,10 +1490,96 @@ class reactor_backend_uring final : public reactor_backend {
+            void complete(size_t fd) noexcept final {
+                _listenfd.speculate_epoll(EPOLLIN);
+                pollable_fd pfd(file_desc::from_fd(fd), pollable_fd::speculation(EPOLLOUT));
+                _result.set_value(std::move(pfd), std::move(_sa));
+                delete this;
+            }
+            void set_exception(std::exception_ptr eptr) noexcept final {
+                _result.set_exception(eptr);
+                delete this;
+            }
+            future<std::tuple<pollable_fd, socket_address>> get_future() {
+                return _result.get_future();
+            }
+            ::sockaddr* posix_sockaddr() {
+                return &_sa.as_posix_sockaddr();
+            }
+            socklen_t* socklen_ptr() {
+                return &_sa.addr_length;
+            }
+        };
+        auto desc = std::make_unique<accept_completion>(listenfd);
+        auto req = internal::io_request::make_accept(listenfd.fd.get(), desc->posix_sockaddr(), desc->socklen_ptr(), SOCK_CLOEXEC);
+        auto fut = desc->get_future();
+        _r._io_sink.submit(desc.release(), std::move(req));
+        return fut.handle_exception_type([&listenfd] (const std::system_error &e) {
+             if (e.code() == std::errc::invalid_argument) {
+                 // The chances are that we shutting down the connection.
+                 listenfd.maybe_no_more_recv();
+             }
+             return make_exception_future<std::tuple<pollable_fd, socket_address>>(e);
+        });
     }
     virtual future<> connect(pollable_fd_state& fd, socket_address& sa) override {
-        return _r.do_connect(fd, sa);
+        class connect_completion final : public io_completion {
+            socket_address _sa;
+            promise<> _result;
+        public:
+            void complete(size_t fd) noexcept final {
+                _result.set_value();
+                delete this;
+            }
+            void set_exception(std::exception_ptr eptr) noexcept final {
+                _result.set_exception(eptr);
+                delete this;
+            }
+            future<> get_future() {
+                return _result.get_future();
+            }
+            ::sockaddr* posix_sockaddr() {
+                return &_sa.as_posix_sockaddr();
+            }
+            socklen_t socklen() const {
+                return _sa.addr_length;
+            }
+        };
+        auto desc = std::make_unique<connect_completion>();
+        auto req = internal::io_request::make_connect(fd.fd.get(), desc->posix_sockaddr(), desc->socklen());
+        auto fut = desc->get_future();
+        _r._io_sink.submit(desc.release(), std::move(req));
+        return fut.handle_exception_type([] (const std::system_error &e) {
+            if (e.code() == std::errc::operation_in_progress) {
+                return make_ready_future();
+            }
+            throw;
+        }).then([&fd, this] {
+            return writeable(fd);
```

Comment:
Why this? It's sure to be writeable after connect().

I think you're confusing it with the async connect protocol where connect() has to be called twice:

```
       EINPROGRESS
              The socket is nonblocking and the connection cannot be completed immediately.  (UNIX domain sockets failed with EAGAIN instead.)  It is possible to select(2) or poll(2) for completion by selecting the socket for
              writing.  After select(2) indicates writability, use getsockopt(2) to read the SO_ERROR option at level SOL_SOCKET to determine whether connect() completed  successfully  (SO_ERROR  is  zero)  or  unsuccessfully
              (SO_ERROR is one of the usual error codes listed here, explaining the reason for the failure).
```

I don't think it's needed here. However, you should speculate the connection as writable immediately after connect succeeds.

Reply:
the fd for connecting is created with `reactor::make_pollable_fd(..., SOCK_NONBLOCK...)`, so it's nonblocking. and the funny thing is that in https://github.com/axboe/liburing/blob/1c469ee303d0f458a6b039c9483865b3af248400/test/connect.c#L135, it does the poll for write and getsockopt() for error dance. and the fd is actually created without `SOCK_NONBLOCK` flag.

but after rereading the kernel source, i still think io_uring retries an io which returns -EAGAIN. i will drop this part and speculate the connection as writable. and retest.

Reply:
Please report it. Either the test code in liburing is wrong (and will confuse others), or we misunderstand something.

Reply:
yup. it's a little bit scaring. i think i will have to prepare a mail to the mailing list along with the "always-retry" finding then. better to be confirmed with liburing's author and the community. i will do so later this week.

Reply:
i just read the issues on the liburing project's GitHub homepage. my "always-retry" understanding was wrong. io_uring  retries only if 

- the op in question is not non-blocking.
- the fd is pollable
- the fd does not have `O_NONBLOCK` flag

see #1248 .